### PR TITLE
Enables setting hatch linewidth in Patches and Collections, also fixes setting hatch linewidth by rcParams

### LINKED ISSUE #21108: [Bug]: Hatch linewidths cannot be modified in an  rcParam context
### Bug summary

The only way (that I am aware of) to control the linewidth of hatches is through an rc parameter. But temporarily modifying the parameter with `plt.rc_context` has not effect.

### Code for reproduction

```python
import matplotlib.pyplot as plt

plt.figure().subplots().bar([0, 1], [1, 2], hatch=["/", "."], fc="r")

with plt.rc_context({"hatch.linewidth": 5}):
    plt.figure().subplots().bar([0, 1], [1, 2], hatch=["/", "."], fc="g")

plt.rc("hatch", linewidth=5)
plt.figure().subplots().bar([0, 1], [1, 2], hatch=["/", "."], fc="b")
```

### Actual outcome

image

image

image

### Expected outcome

That second image (the green bars) should have thick hatches.

FWIW I think hatches ought to have an actual API, but given that they don't, this limitation makes them *really* hard to work with.

### Operating system

macos

### Matplotlib Version

3.4.3

### Matplotlib Backend

inline

### PR DESCRIPTION
**Pull request by Impaler343**
## PR summary

Fixes #21108 using the patch provided by @anntzer which I modified to pass tests. Need to write tests for:

- [x] Setting hatch linewidth in a Patch
- [x] Setting hatch linewidth in a Collection

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by Impaler343** [comment]
I don't know why docs and CI are failing, can somebody check this out?

**Comment by rcomer** [comment]
Hi @Impaler343, I'm not certain but I think the docs failure may be related to something that has been fixed on `main`.  Could you try rebasing?

```
git fetch upstream
git rebase upstream/main
```

Appveyor is showing the same failure on all PRs right now, so I wouldn't worry about that one.

**Comment by Impaler343** [comment]
Do I also need to squash?

File: lib/matplotlib/tests/test_collections.py [hunk 1]
```diff
@@ -1336,3 +1336,24 @@ def test_striped_lines(fig_test, fig_ref, gapcolor):
     for x, gcol, ls in zip(x, itertools.cycle(gapcolor),
                            itertools.cycle(linestyles)):
         ax_ref.axvline(x, 0, 1, linestyle=ls, gapcolor=gcol, alpha=0.5)
+
+
+@check_figures_equal(extensions=['png', 'pdf', 'svg', 'eps'])
+def test_hatch_linewidth(fig_test, fig_ref):
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_collections.py:1342]: This test does not seem correct.  Looking at what this code draws there is nothing shown (I think all of the circles are drawn at 0, 0 in screen space and clipped to not be visible by the axes).

File: lib/matplotlib/hatch.py [hunk 2]
```diff
@@ -223,3 +223,14 @@ def get_path(hatchpattern, density=6):
             cursor += pattern.num_vertices
 
     return Path(vertices, codes)
+
+
+def _set_hatch_linewidth(gc, hatch_linewidth):
```
Comment (tacaswell) [inline lib/matplotlib/hatch.py:228]: This probably is a better fit in backend bases?  Even though it is just for hatches, it is really about functionality of the graphics context objects.

File: lib/matplotlib/backends/backend_ps.py [hunk 3]
```diff
@@ -344,12 +344,12 @@ def set_font(self, fontname, fontsize, store=True):
                 self.fontname = fontname
                 self.fontsize = fontsize
 
-    def create_hatch(self, hatch):
+    def create_hatch(self, gc, hatch):
```
Comment (ksunden) [inline lib/matplotlib/backends/backend_ps.py:347]: Should probably keep `gc` as the second argument here. Possibly even with fallback to the rcparam if `gc=None` (and thus have `None` as the default)

While it would be a strange thing to call outside of the class, `create_hatch` is not marked as private, and so is technically a public API, so I would err on the side of caution in changing the signature. (i.e. adding an optional parameter is okay, but adding a required parameter or changing order would technically require a deprecation cycle... this is on the edge of being okay to change since it is not likely used many places, but given that it is easy to make it fully back-compatible, inclined to do so)

File: lib/matplotlib/backends/backend_ps.py [hunk 4]
```diff
@@ -361,9 +361,9 @@ def create_hatch(self, hatch):
 
      /PaintProc {{
         pop
-        {linewidth:g} setlinewidth
-{self._convert_path(
-    Path.hatch(hatch), Affine2D().scale(sidelen), simplify=False)}
+        {linewidth:f} setlinewidth
+        {self._convert_path(
+            Path.hatch(hatch), Affine2D().scale(sidelen), simplify=False)}
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_ps.py:366]: Can you back these changes all the way out?

File: lib/matplotlib/backends/backend_pdf.py [hunk 5]
```diff
@@ -2508,14 +2512,15 @@ def alpha_cmd(self, alpha, forced, effective_alphas):
...
-    def hatch_cmd(self, hatch, hatch_color):
+    def hatch_cmd(self, hatch, hatch_color, hatch_linewidth):
         if not hatch:
             if self._fillcolor is not None:
                 return self.fillcolor_cmd(self._fillcolor)
             else:
                 return [Name('DeviceRGB'), Op.setcolorspace_nonstroke]
         else:
-            hatch_style = (hatch_color, self._fillcolor, hatch)
+            hatch_style = (hatch_color, self._fillcolor, hatch,
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_pdf.py:2522]: picky: this can be one line.

**Comment by tacaswell** [comment]
There is enough exploration/experimentation in here it is probably worth squashing to a handful of commits.

**Comment by tacaswell** [comment]
I think this is like 80% of the way there and looking pretty good!

File: lib/matplotlib/hatch.py (hunk 2 above)
Comment (Impaler343) [reply lib/matplotlib/hatch.py:228]: Hmmmm, makes sense

File: lib/matplotlib/backends/backend_ps.py (hunk 4 above)
Comment (Impaler343) [reply lib/matplotlib/backends/backend_ps.py:366]: These changes were done as flake8 failed while committing

File: lib/matplotlib/backends/backend_ps.py (hunk 3 above)
Comment (Impaler343) [reply lib/matplotlib/backends/backend_ps.py:347]: As it isn't usually called outside of the class, why don't we make it private? With the changes you mentioned. 
Or essentially do you want it to work without passing gc to the function?

File: lib/matplotlib/backends/backend_ps.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/backends/backend_ps.py:366]: flake8 should not have flagged either the g->f change or indentation in a multi-line string.  

If flake8 is indeed triggering and requiring either of these changes that is a bug is flake8.

Comment (Impaler343) [reply lib/matplotlib/backends/backend_ps.py]: I've committing after skipping flake8, can't do anything else I think

**Comment by Impaler343** [comment]
Can we get a feature for ```check_figures_equal()``` similar to ```image_comparison()```'s  ```remove_text``` feature. As the former is a more economical way to test our features, having it for ```check_figures_equal()``` would be a good idea. Or is it already available, and I don't see it? In which case I suggest we update the docs

**Comment by oscargus** [comment]
> Can we get a feature for `check_figures_equal()` similar to `image_comparison()`'s `remove_text` feature? 

I do not think this is required? The reason `remove_text` exists is that the font files may be different on different machine. However, `check_figures_equal()` are executed on the same machine, so should never be a problem. Am I missing something?

**Comment by Impaler343** [comment]
> > Can we get a feature for `check_figures_equal()` similar to `image_comparison()`'s `remove_text` feature?
> 
> I do not think this is required? The reason `remove_text` exists is that the font files may be different on different machine. However, `check_figures_equal()` are executed on the same machine, so should never be a problem. Am I missing something?

I suggested this as it is a quick plugin to make tests run faster by focussing on the important part of the plot, by removing axes, labels and ticks, and I think that this will make a significant difference while running a large amount of tests

File: lib/matplotlib/backend_bases.py [hunk 6]
```diff
@@ -1006,6 +1006,20 @@ def get_hatch_linewidth(self):
...
 
+    def set_hatch_linewidth(self, hatch_linewidth):
+        """Set the hatch linewidth."""
+        self._hatch_linewidth = hatch_linewidth
+
+    def _set_hatch_linewidth(self, hatch_linewidth):
+        if hasattr(self, "set_hatch_linewidth"):
+            self.set_hatch_linewidth(hatch_linewidth)
+        else:
+            _api.warn_deprecated(
```
Comment (ksunden) [inline lib/matplotlib/backend_bases.py:1017]: Why is this warning here? 

You define the method that you are `hasattr` checking in the same class as the `_` method, so if the `_` method exists, the public version will exist on `self`. 


Additionally, if there is a warning here, (which I'm not sure a deprecation makes sense here), the removal would be 3.10, not 3.8 (since we already branched for 3.9)

Comment (Impaler343) [reply lib/matplotlib/backend_bases.py:1017]: I put this in initially thinking there wouldn't be immediate support for all backends. You are right, I think a warning is not necessary. I'll just put in a message like in ```set_alpha``` saying it is not supported in all backends yet. Hatches don't render on wxcairo and pgf.

File: lib/matplotlib/backends/backend_pdf.py [hunk 7]
```diff
@@ -1535,25 +1535,29 @@ def _write_soft_mask_groups(self):
...
+        pattern = self._hatch_patterns.get(hatch_style, None)
         if pattern is not None:
             return pattern
 
         name = next(self._hatch_pattern_seq)
-        self.hatchPatterns[hatch_style] = name
+        self._hatch_patterns[hatch_style] = name
         return name
 
+    hatchPatterns = _api.deprecated("3.8")(property(lambda self: {
```
Comment (oscargus) [inline lib/matplotlib/backends/backend_pdf.py:1553]:
```suggestion
    hatchPatterns = _api.deprecated("3.10")(property(lambda self: {
```

An API change note is also required. See https://matplotlib.org/devdocs/devel/api_changes.html#deprecate-api

Comment (Impaler343) [reply lib/matplotlib/backends/backend_pdf.py]: Could you please check if the API change note is written correctly? Also the failing test is unrelated

**Comment by Impaler343** [comment]
Have removed the if conditions as they are unreachable:
 - ```hatchstyle``` cannot be None as it is checked for in the function ```hatch_cmd``` before being called.
 - ```edge``` always receives a numpy array of (r, g, b, alpha) values
 
On another note, a lot of lines in backend_bases is uncovered, and I don't think I can write tests for all of them. What to do?

File: doc/api/next_api_changes/deprecations/28048-PR.rst [hunk 8]
```diff
@@ -0,0 +1,6 @@
+``PdfFile.hatchPatterns``
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+... is deprecated and replaced by an internal helper called ``PdfFile._hatch_patterns`` instead.
+
+- The attribute ``PdfFile._hatch_patterns`` now stores hatch linewidths as well.
```
Comment (QuLogic) [inline doc/api/next_api_changes/deprecations/28048-PR.rst:6]: There is no need to mention the private versions.

File: lib/matplotlib/backend_bases.py [hunk 9]
```diff
@@ -1006,6 +1006,10 @@ def get_hatch_linewidth(self):
         """Get the hatch linewidth."""
         return self._hatch_linewidth
 
+    def set_hatch_linewidth(self, hatch_linewidth):
+        """Set the hatch linewidth - not supported on all backends."""
```
Comment (QuLogic) [inline lib/matplotlib/backend_bases.py:1010]: Which ones are not supported?

File: lib/matplotlib/backends/backend_pdf.py [hunk 10]
```diff
@@ -2460,7 +2463,7 @@ def stroke(self):
         """
         # _linewidth > 0: in pdf a line of width 0 is drawn at minimum
         #   possible device width, but e.g., agg doesn't draw at all
-        return (self._linewidth > 0 and self._alpha > 0 and
+        return (self.get_linewidth() > 0 and self._alpha > 0 and
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:2466]: Is this change related? I don't see any change to `get_linewidth` or `_linewidth` that might require it.

File: lib/matplotlib/backends/backend_ps.py [hunk 11]
```diff
@@ -361,8 +361,9 @@ def create_hatch(self, hatch):
 
      /PaintProc {{
         pop
-        {linewidth:g} setlinewidth
-{self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen), simplify=False)}
+        {linewidth:f} setlinewidth
+        {self._convert_path(
+            Path.hatch(hatch), Affine2D().scale(sidelen), simplify=False)}
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py:366]: Unrelated change:
```suggestion
        {linewidth:g} setlinewidth
{self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen), simplify=False)}
```

File: lib/matplotlib/collections.py [hunk 12]
```diff
@@ -541,6 +543,14 @@ def get_hatch(self):
         """Return the current hatching pattern."""
         return self._hatch
 
+    def set_hatch_linewidth(self, lw):
+        """Set the hatch linewidth"""
+        self._hatch_linewidth = lw
+
+    def get_hatch_linewidth(self):
+        """Return the hatch linewidth"""
+        return self._hatch_linewidth
```
Comment (QuLogic) [inline lib/matplotlib/collections.py:552]:
```suggestion
        """Set the hatch linewidth."""
        self._hatch_linewidth = lw

    def get_hatch_linewidth(self):
        """Return the hatch linewidth."""
        return self._hatch_linewidth
```

File: lib/matplotlib/patches.py [hunk 13]
```diff
@@ -571,6 +572,14 @@ def get_hatch(self):
         """Return the hatching pattern."""
         return self._hatch
 
+    def set_hatch_linewidth(self, hatch_linewidth):
+        """Set the hatch linewidth"""
+        self._hatch_linewidth = hatch_linewidth
+
+    def get_hatch_linewidth(self):
+        """Return the hatch linewidth"""
+        return self._hatch_linewidth
```
Comment (QuLogic) [inline lib/matplotlib/patches.py:581]: Probably should be consistent with the `Collection` version:
```suggestion
    def set_hatch_linewidth(self, lw):
        """Set the hatch linewidth."""
        self._hatch_linewidth = lw

    def get_hatch_linewidth(self):
        """Return the hatch linewidth."""
        return self._hatch_linewidth
```

File: lib/matplotlib/patches.pyi [hunk 14]
```diff
@@ -59,6 +59,8 @@ class Patch(artist.Artist):
     def set_joinstyle(self, s: JoinStyleType) -> None: ...
     def get_joinstyle(self) -> Literal["miter", "round", "bevel"]: ...
     def set_hatch(self, hatch: str) -> None: ...
+    def set_hatch_linewidth(self, hatch_linewidth: float) -> None: ...
```
Comment (QuLogic) [inline lib/matplotlib/patches.pyi:62]: Ditto.
```suggestion
    def set_hatch_linewidth(self, lw: float) -> None: ...
```

File: lib/matplotlib/tests/test_collections.py [hunk 15]
```diff
@@ -1336,3 +1336,27 @@ def test_striped_lines(fig_test, fig_ref, gapcolor):
...
+    polygons = [
+        [(0.1, 0.1), (0.1, 0.4), (0.4, 0.4), (0.4, 0.1)],
+        [(0.6, 0.6), (0.6, 0.9), (0.9, 0.9), (0.9, 0.6)]
+    ]
+    ref = PolyCollection(polygons, hatch="x")
+    ref.set_hatch_linewidth(lw)
+
+    with mpl.rc_context({"hatch.linewidth": lw}):
+        test = PolyCollection(polygons, hatch="x")
+        test.set_hatch_linewidth(lw)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_collections.py:1357]: Why set the `rc_context` and call `set_hatch_linewidth`?

File: lib/matplotlib/tests/test_collections.py [hunk 16]
```diff
@@ -1336,3 +1336,27 @@ def test_striped_lines(fig_test, fig_ref, gapcolor):
...
+
+    ax_ref.add_collection(ref)
+    ax_test.add_collection(test)
+
+    assert (test.get_hatch_linewidth() == lw) and (ref.get_hatch_linewidth() == lw)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_collections.py:1362]:
```suggestion
    assert test.get_hatch_linewidth() == ref.get_hatch_linewidth() == lw
```

File: lib/matplotlib/tests/test_patches.py [hunk 17]
```diff
@@ -960,3 +960,20 @@ def test_arrow_set_data():
...
+    lw = 2.0
+
+    with plt.rc_context({"hatch.linewidth": lw}):
+        ax_ref.add_patch(mpatches.Rectangle((0, 0), 1, 1, hatch="x"))
+
+    ax_test.add_patch(mpatches.Rectangle((0, 0), 1, 1, hatch="x"))
+    ax_test.patches[0].set_hatch_linewidth(lw)
+
+    assert (ax_ref.patches[0].get_hatch_linewidth() == lw)
+    assert (ax_test.patches[0].get_hatch_linewidth() == lw)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_patches.py:979]:
```suggestion
    assert ax_ref.patches[0].get_hatch_linewidth() == lw
    assert ax_test.patches[0].get_hatch_linewidth() == lw
```

**Review by QuLogic — COMMENTED** [review]
It appears this needs a rebase, but is mostly complete.

File: lib/matplotlib/backends/backend_pdf.py (hunk 10 above)
Comment (Impaler343) [reply lib/matplotlib/backends/backend_pdf.py:2466]: Not really, I just felt accessing a private variable using the _getter_ would be more appropriate

File: lib/matplotlib/backend_bases.py (hunk 9 above)
Comment (Impaler343) [reply lib/matplotlib/backend_bases.py:1010]: Was facing some issues with Cairo backends but not anymore

File: lib/matplotlib/tests/test_collections.py (hunk 15 above)
Comment (Impaler343) [reply lib/matplotlib/tests/test_collections.py]: Missed that :sweat_smile:

File: lib/matplotlib/backends/backend_svg.py [hunk 18]
```diff
@@ -500,11 +500,12 @@ def _get_hatch(self, gc, rgbFace):
         edge = gc.get_hatch_color()
         if edge is not None:
             edge = tuple(edge)
-        dictkey = (gc.get_hatch(), rgbFace, edge)
+        lw = gc._linewidth
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py:503]: Should be hatch, not regular, line width:
```suggestion
        lw = gc.get_hatch_linewidth()
```
which should fix the tests.

**Comment by QuLogic** [comment]
I _think_ the doc build failure is unrelated; you may have to rebase/merge to fix it.

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/backends/backend_ps.py [hunk 19]
```diff
@@ -505,12 +505,12 @@ def set_font(self, fontname, fontsize, store=True):
                 self.fontname = fontname
                 self.fontsize = fontsize
 
-    def create_hatch(self, hatch):
+    def create_hatch(self, hatch, gc=None):
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_ps.py:508]:
```suggestion
    def create_hatch(self, hatch, gc):
```

If `gc` is None this will fail so make it required.

Comment (tacaswell) [inline lib/matplotlib/backends/backend_ps.py:508]:
```suggestion
    def create_hatch(self, hatch, lw):
```

File: lib/matplotlib/backends/backend_ps.py [hunk 20]
```diff
@@ -933,7 +933,7 @@ def _draw_ps(self, ps, gc, rgbFace, *, fill=True, stroke=True):
                 write("grestore\n")
 
         if hatch:
-            hatch_name = self.create_hatch(hatch)
+            hatch_name = self.create_hatch(hatch, gc)
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_ps.py:936]:
```suggestion
            hatch_name = self.create_hatch(hatch, gc.get_hatch_linewidth())
```

File: lib/matplotlib/backends/backend_ps.py [hunk 21]
```diff
@@ -505,12 +505,12 @@ def set_font(self, fontname, fontsize, store=True):
...
                 self.fontsize = fontsize
 
-    def create_hatch(self, hatch):
+    def create_hatch(self, hatch, gc=None):
         sidelen = 72
         if hatch in self._hatches:
             return self._hatches[hatch]
         name = 'H%d' % len(self._hatches)
-        linewidth = mpl.rcParams['hatch.linewidth']
+        linewidth = gc.get_hatch_linewidth()
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_ps.py:513]:
```suggestion
        linewidth = lw
```

File: lib/matplotlib/backends/backend_ps.py (hunk 19 above)
Comment (tacaswell) [reply lib/matplotlib/backends/backend_ps.py:508]: I suggest either this or the other three suggestions (pass in lw like the other backends rather than the whole gc).

Comment (Impaler343) [reply lib/matplotlib/backends/backend_ps.py]: Passing in only linewidth seems like a good idea

File: lib/matplotlib/backends/backend_ps.py [hunk 22]
```diff
@@ -505,12 +505,12 @@ def set_font(self, fontname, fontsize, store=True):
...
                 self.fontsize = fontsize
 
-    def create_hatch(self, hatch):
+    def create_hatch(self, hatch, lw):
         sidelen = 72
         if hatch in self._hatches:
             return self._hatches[hatch]
         name = 'H%d' % len(self._hatches)
-        linewidth = mpl.rcParams['hatch.linewidth']
+        linewidth = lw
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_ps.py:513]: Might as well drop the extra variable, though:
```suggestion
    def create_hatch(self, hatch, linewidth):
        sidelen = 72
        if hatch in self._hatches:
            return self._hatches[hatch]
        name = 'H%d' % len(self._hatches)
```
